File: js/src/bowlingGame/infrastructure/reportingTest.js
import * as Tests from "../bowlingGame.test";
import ResultReporter from "./resultReporter";
import jsonfile from "jsonfile";
import fs from "fs";

let currentRunTests = [];
let now = new Date().toISOString();
const resultsFileName = "results.json";

describe("Bowling game", () => {
    Tests.default();

    before(function () {
        if (Tests.Names === "ENTER YOUR NAME HERE") {
            throw new Error("Please enter your name in bowlingGame.test.js");
        }
        if (fs.existsSync(resultsFileName))
            currentRunTests = jsonfile.readFileSync(resultsFileName);;
    });
    after(function () {
        var reporter = new ResultReporter();
        currentRunTests = currentRunTests.filter(x => x.LastRunTime === now);
        reporter.reportResults(currentRunTests, Tests.Names);
    });
    beforeEach(function () {

    });
    afterEach(function () {
        let testName = this.currentTest.fullTitle();
        let foundTest = currentRunTests.find(x => x.TestName === testName);
        if (foundTest) {
            foundTest.LastRunTime = now;
            foundTest.Succeeded = this.currentTest.state === "passed";
            return;
        }
        currentRunTests.push(
            {
                FirstRunTime: now,
                LastRunTime: now,
                TestName: testName,
                TestMethod: this.currentTest.title,
                Succeeded: this.currentTest.state === "passed"
            });
        jsonfile.writeFileSync(resultsFileName, currentRunTests);
    });
});

File: cs/BowlingGame/Infrastructure/ReportingTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Newtonsoft.Json;
using NUnit.Framework;
using NUnit.Framework.Interfaces;

namespace BowlingGame.Infrastructure
{
    public class ReportingTest<TTestClass>
    {
        private static readonly string resultsFileName = typeof(TTestClass).Name + ".json";
        private static string resultsFile;
        private static List<TestCaseStatus> tests;

        [OneTimeSetUp]
        public void ClearLocalResults()
        {
            resultsFile = Path.Combine(TestContext.CurrentContext.TestDirectory, resultsFileName);
            tests = LoadResults();
        }

        [TearDown]
        public static void WriteLastRunResult()
        {
            var test = TestContext.CurrentContext.Test;
            var status = TestContext.CurrentContext.Result.Outcome.Status;
            var succeeded = status == TestStatus.Passed;

            var testName = test.Name;
            if (!test.Name.Contains(test.MethodName))
                testName = test.MethodName + " " + test.Name;

            var testStatus = tests.FirstOrDefault(t => t.TestName == testName);
            if (testStatus != null)
            {
                testStatus.LastRunTime = DateTime.Now;
                testStatus.Succeeded = succeeded;
            }
            else
                tests.Add(new TestCaseStatus
                {
                    FirstRunTime = DateTime.Now,
                    LastRunTime = DateTime.Now,
                    TestName = testName,
                    TestMethod = test.MethodName,
                    Succeeded = succeeded
                });
        }

        private static void SaveResults(List<TestCaseStatus> tests)
        {
            File.WriteAllText(resultsFile, JsonConvert.SerializeObject(tests, Formatting.Indented));
        }

        private static List<TestCaseStatus> LoadResults()
        {
            try
            {
                var json = File.ReadAllText(resultsFile);
                var statuses = JsonConvert.DeserializeObject<List<TestCaseStatus>>(json);
                return RemoveOldNames(statuses);
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return new List<TestCaseStatus>();
            }
        }

        private static List<TestCaseStatus> RemoveOldNames(List<TestCaseStatus> statuses)
        {
            var names = new HashSet<string>(typeof(TTestClass).GetMethods().Select(m => m.Name));
            return statuses.Where(s => names.Contains(s.TestMethod)).ToList();
        }

        [OneTimeTearDown]
        public static void ReportResults()
        {
            tests = tests.OrderByDescending(t => t.LastRunTime).ThenByDescending(t => t.FirstRunTime).ToList();
            SaveResults(tests);

            foreach (var kv in tests)
                Console.WriteLine(kv.TestName);

            if (string.IsNullOrWhiteSpace(YourName.Authors))
                throw new Exception("Enter your surnames at YourName.cs in AUTHORS constant");

            using (var client = Firebase.CreateClient())
            {
                string authorsKey = MakeFirebaseSafe(YourName.Authors);
                client.Set(authorsKey, new
                {
                    tests,
                    time = DateTime.Now.ToUniversalTime(),
                    lang = "cs"
                });
            }
            Console.WriteLine("reported");
        }

        private static string MakeFirebaseSafe(string s)
        {
            var badChars = Enumerable.Range(0, 32).Select(code => (char)code)
                .Concat(".$#[]/" + (char)127);
            foreach (var badChar in badChars)
                s = s.Replace(badChar, '_');
            return s;
        }
    }
}